### PR TITLE
feat: Added IAMSupport for Postgres

### PR DESCRIPTION
**Pull request by ClarLabs**
This is a re-release of PR #1514 with minor changes.

### CODE REVIEW

**Review by not-night-but — CHANGES_REQUESTED** [review]
Hey Chris!

This looks great, just left one nitpick and a question.

Let me know if you have any questions/gripes

File: apps/studio/package.json [hunk 1]
```diff
@@ -24,6 +23,8 @@
   "main": "dist/background.js",
   "dependencies": {
     "@aws-sdk/client-redshift": "^3.145.0",
+    "@aws-sdk/credential-providers": "^3.614.0",
```
Comment (rathboma) [inline apps/studio/package.json:26]: all aws-sdk libs should probably be the same version

File: apps/studio/src/components/connection/PostgresForm.vue [hunk 2]
```diff
@@ -16,7 +13,54 @@
         type="text"
         class="form-control"
         v-model="config.options.cluster"
+      />
+    </div>
+    <common-advanced :config="config" />
+
+    <div class="advanced-connection-settings">
```
Comment (rathboma) [inline apps/studio/src/components/connection/PostgresForm.vue:20]: is this only for Redshift? Does it work for RDS Postgres also, or is that out of scope of this PR?

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 3]
```diff
@@ -1223,10 +1225,34 @@ export class PostgresClient extends BasicDatabaseClient<QueryResult> {
   }
 
   protected async configDatabase(server: IDbConnectionServer, database: { database: string}) {
+
+    let resolvedPw = null;
+    const redshiftOptions = server.config.redshiftOptions;
+    console.log("redshiftOptions", redshiftOptions);
+    if (
```
Comment (rathboma) [inline apps/studio/src/lib/db/clients/postgresql.ts:1232]: Seems like this change only adds support for credentials stored in a credentials file, is that correct?

**Review by rathboma — COMMENTED** [review]
Looks good. This is the IAM branch from previously it looks like?

Added a couple of questions / minor comments

**Comment by rathboma** [comment]
TODO before this can be merged:

1. Needs adding to the docs (with screenshots)

File: apps/studio/src/components/connection/PostgresForm.vue [hunk 4]
```diff
@@ -16,7 +13,54 @@
...
+        <div class="form-group">
+          <label for="awsProfile"> AWS Profile </label>
+          <input
+            name="awsProfile"
+            type="text"
+            class="form-control"
+            v-model="config.redshiftOptions.awsProfile"
+          />
+        </div>
+      </div>
```
[the remark this answers is not in the document]
Comment (rathboma) [reply apps/studio/src/components/connection/PostgresForm.vue:63]: Good point.

We'll need to support other ways to use Iam pretty soon - including running a script to get IAM credentials

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 5]
```diff
@@ -1223,10 +1225,34 @@ export class PostgresClient extends BasicDatabaseClient<QueryResult> {
...
+      });
+      const signer = new Signer({
+        credentials: nodeProviderChainCredentials,
+        region: redshiftOptions?.awsRegion,
+        hostname: server.config.host,
+        port: server.config.port,
+        username: server.config.user,
+      });
+
+      console.log(server.config.host, server.config.port, server.config.user)
```
[the remark this answers is not in the document]
Comment (ClarLabs) [reply apps/studio/src/lib/db/clients/postgresql.ts:1247]: Ah my bad, will remove that

File: apps/studio/src/components/connection/PostgresForm.vue (hunk 2 above)
Comment (ClarLabs) [reply apps/studio/src/components/connection/PostgresForm.vue:20]: I tested on RDS Postgres

File: apps/studio/src/lib/db/clients/postgresql.ts (hunk 3 above)
Comment (ClarLabs) [reply apps/studio/src/lib/db/clients/postgresql.ts:1232]: That is correct, I need to investigate the CLI auth, that WILL be part of this PR unless you want it in a seperate PR, but if it's anything like Azure CLI, I can't see it being a huge change.

**Comment by rathboma** [comment]
@ClarLabs hey! I want to get this merged -- looks like you'll need to merge master into your branch, and add a walkthrough to the docs. Then it should be good to go.

**Comment by ClarLabs** [comment]
> @ClarLabs hey! I want to get this merged -- looks like you'll need to merge master into your branch, and add a walkthrough to the docs. Then it should be good to go.

Hey Matt, I will get this done over the weekend and let you know once complete

File: docs/user_guide/connecting/amazon-rdb.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
+Connecting to Amazon RDB requires you to set up an IAM user and ensure your security group allows traffic from your IP address.
+
+## IAM Configuration in Amazon
```
Comment (rathboma) [inline docs/user_guide/connecting/amazon-rdb.md:3]: Questions:

- [ ] This is just for postgres, no?
- [ ] Does this require them to have a credential file locally? If so, needs mentioning, and a link providing to how to set that up
- [ ] Are there any links you can provide to AWS docs for how to do some of this stuff, eg setting up an IAM user, opening ports in security groups?

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 7]
```diff
@@ -1257,11 +1259,33 @@ export class PostgresClient extends BasicDatabaseClient<QueryResult> {
...
+    const redshiftOptions = server.config.redshiftOptions;
+
+    if (
+      server.config.client === "postgresql" &&
+      redshiftOptions?.iamAuthenticationEnabled
+    ) {
+      const nodeProviderChainCredentials = fromIni({
+        profile: redshiftOptions.awsProfile ?? "default",
+      });
+      const signer = new Signer({
```
Comment (rathboma) [inline apps/studio/src/lib/db/clients/postgresql.ts:1274]: Question: Does this work with SSH tunnels?

**Review by rathboma — COMMENTED** [review]
Mostly docs questions. Also -- does this work with an SSH tunnel? Does SSL always need to be enabled if it's IAM authenticated?

File: apps/studio/src/lib/db/clients/postgresql.ts (hunk 7 above)
Comment (ClarLabs) [reply apps/studio/src/lib/db/clients/postgresql.ts:1274]: This does work via an SSH Tunnel, I tested locally ;)

File: docs/user_guide/connecting/amazon-rdb.md (hunk 6 above)
Comment (ClarLabs) [reply docs/user_guide/connecting/amazon-rdb.md:3]: This has only be done on Postgres, However, if you give me one more week, I can see about Aurua DB, Maria DB, and MSQL Server or can do it as another PR.

This does still require credential file locally - updated the docs

Updated the docs to include relevant links.

**Comment by rathboma** [comment]
@ClarLabs some video notes on the UI:

https://github.com/user-attachments/assets/b9fba695-348a-4b16-a3cf-d39cd415002f

**Comment by ClarLabs** [comment]
> @ClarLabs some video notes on the UI:
> 
>  720-2024-09-14.11-53-15.mp4

Hey Matt, thanks for the feedback, I have made some changes but need further testing, I should have it completed by tomorrow.

FYI, the SSL tunnel still works with it, I tested it on a SSH box here and it worked as expected and failed when I wanted it too (wrong SSH details etc)

Turns out port is required during the auth process so it's default if not specified, but changed on AWS it can be changed here too

**Comment by rathboma** [comment]
@ClarLabs sounds good thanks! Please let me know when it's ready for final review.

**Comment by ClarLabs** [comment]
Sorry, meant to update this, it's ready for it's final review, everything seems to work, just needed to tweak message.

File: apps/studio/src/lib/db/clients/postgresql.ts (hunk 3 above)
Comment (rathboma) [reply apps/studio/src/lib/db/clients/postgresql.ts:1267]: Let's do a separate PR

**Review by rathboma — APPROVED** [review]

**Pull request merged**